jog_controller/src/twist_to_jog_frame.cpp

```diff
@@ -0,0 +1,193 @@
+#include "jog_msgs/GetTargetList.h"
+#include "jog_msgs/SetTarget.h"
+#include "jog_msgs/SetCollisionAvoidance.h"
+
+namespace jog_controller
+{
+
+TwistToJogFrame::TwistToJogFrame()
+{
+  ros::NodeHandle nh, pnh("~");
+  jog_frame_pub_ = nh.advertise<jog_msgs::JogFrame>("jog_frame", 1);
+  get_target_frame_list_srv_ = nh.advertiseService("/twist_to_jog_frame/get_target_frame_list", &TwistToJogFrame::getTargetFrameList, this);
+  set_controller_status_srv_ = nh.advertiseService("/twist_to_jog_frame/set_controller_status", &TwistToJogFrame::setControllerStatus, this);
+  set_target_frame_srv_ = nh.advertiseService("/twist_to_jog_frame/set_target_frame", &TwistToJogFrame::setTargetFrame, this);
+  set_target_link_srv_ = nh.advertiseService("/twist_to_jog_frame/set_target_link", &TwistToJogFrame::setTargetLink, this);
+  pnh.getParam("/jog_frame_node/group_names", group_names_);
+  pnh.getParam("/jog_frame_node/link_names", link_names_);
+  pnh.getParam("group_name", group_name_);
+  pnh.getParam("link_name", link_name_);
+  pnh.getParam("frame_id", frame_id_);
+  pnh.getParam("rotate_axes", rotate_axes_);
+  std::string str_rotation_matrix_;
+  pnh.getParam("rotation_matrix", str_rotation_matrix_); 
+  rotation_matrix_ = arma::mat(str_rotation_matrix_);
+  pnh.getParam("dominant_axis_mode", dominant_axis_mode_);
+  pnh.getParam("scale_linear", scale_linear_);
+  pnh.getParam("scale_angular", scale_angular_);
+  pnh.getParam("sub_topic", sub_topic_);
+  pnh.param("avoid_collisions", avoid_collisions_, true);
+  pnh.getParam("controller_enabled", controller_enabled_);
+  twist_sub_ = nh.subscribe(sub_topic_, 10, &TwistToJogFrame::twist_cb, this);
+  ros::topic::waitForMessage<geometry_msgs::Twist>(sub_topic_);
+}
+
+void TwistToJogFrame::keepOnlyDominantAxis(arma::vec6 &twist_p)
+{
+  double min = twist_p.min();
+  double max = twist_p.max();
+  double min_max = max + min;
+  if(min_max > 0){
+    arma::uword i = twist_p.index_max();
+    twist_p.zeros();
+    twist_p.at(i) = max;
+  } else {
+    arma::uword i = twist_p.index_min();
+    twist_p.zeros();
+    twist_p.at(i) = min;
+  }
+}
+
+void TwistToJogFrame::rotateAxes(arma::vec6 &twist_p)
+{
+  arma::vec linear = arma::vec3({twist_p.at(0), twist_p.at(1), twist_p.at(2)});
+  arma::vec angular = arma::vec3({twist_p.at(3), twist_p.at(4), twist_p.at(5)});
+
+  linear = rotation_matrix_ * linear;
+  angular = rotation_matrix_ * angular;
+
+  //Write in original vector
+  for(arma::uword i = 0; i < 3; i++){ 
+    twist_p[i] = linear[i];
+    twist_p[i+3] = angular[i];
+  }
+}
+
+bool TwistToJogFrame::hasDuplicate(const arma::vec6 &twist_p)
+{
+  arma::vec nz_tmp = arma::nonzeros(twist_p);
+  arma::vec unique_tmp = arma::unique(nz_tmp);
+  return size(unique_tmp) != size(nz_tmp);
+}
+
+void TwistToJogFrame::scaleCommand(arma::vec6 &twist_p)
+{
+  for(arma::uword i = 0; i < 3; i++)
+  {
+    twist_p.at(i) = twist_p.at(i) * scale_linear_;
+  }
+  for(arma::uword i = 3; i < 6; i++)
```

Comment:
you could combine both for and offset `i` by `+3` on each loop for the scale_angular